### PR TITLE
West runners dependencies in sync with CMake runners target 

### PR DESCRIPTION
**Pull request by tejlmand**
This commit creates a phony west_flash_target target with identical
dependencies as CMake flash target (and the same for debug, debugserver,
etc. targets).

ninja flash correctly ensures dependencies are taken into
consideration before calling west flash.

Unfortunately, calling west flash directly would not re-run the flash
dependencies, cause west flash would only build the default CMake
target.

Now, west flash calls the phony west_flash_depends target, ensuring
all deps are up-to-date before flashing (unless --skip-rebuild is given)

--------

Manifest: https://github.com/nrfconnect/sdk-nrf/pull/2499

### CODE REVIEW

File: scripts/west_commands/run_common.py [hunk 1]
```diff
@@ -142,90 +142,91 @@ def do_run_common(command, user_args, user_runner_args):
...
-        log.die('required program', e.filename,
-                'not found; install it or add its location to PATH')
-    except RuntimeError as re:
-        if not user_args.verbose:
-            log.die(re)
-        else:
-            log.err('verbose mode enabled, dumping stack:', fatal=True)
-            raise
+    for yaml in runners_yaml:
+        runner_config = load_runners_yaml(yaml, user_args)
```
Comment (tejlmand) [inline scripts/west_commands/run_common.py:148]: Note: No code changes, only indentation due to the for loop modification.

**Comment by mbolivar-nordic** [comment]
Can you explain why this changed to multiple files, instead of changing the data within the file to be more complex as needed?

**Comment by tejlmand** [comment]
> Can you explain why this changed to multiple files, instead of changing the data within the file to be more complex as needed?

because this way we keep the logic completely separated.
In this case, the subimage `hci_rpmsg` creates it's own runner, and the parent image does not need to know or parse what the content of that runners.yam file is.

The subimage can simply specifies that it produces a `runners.yaml` as a byproduct.
So think of it as similar to CMake External project/subimage, where the parent CMake build system does not need to know anything about the internals of the build of the child image.

I don't see the value in added complexity of merging the two (or more) runners files.
That feels like similar to old `${IMAGE}` in CMake, where parent build system would know everything about child image.

**Comment by mbolivar-nordic** [comment]
OK, I'm starting to feel a bit lost in all this; could we have a quick call on Monday morning my time / afternoon your time to discuss the problems and solutions? @oyvindronningstad perhaps you could join too?

**Comment by tejlmand** [comment]
> 
> OK, I'm starting to feel a bit lost in all this; could we have a quick call on Monday morning my time / afternoon your time to discuss the problems and solutions? @oyvindronningstad perhaps you could join too?

Ahh, now I get it, you're talking about this solution compared to #322.
(I though you were suggesting to extend the info in `runners.yml`)

The  reason for my proposal is to use the info we already have.
Solution in #322 would do the following:
1) Build `child_image.hex`.
    - and create a child_image/runners.yml
      (each image always create a runners.yml)
2) Build `parent_image.hex`
    - and create a parent_image/runners.yml
      (each image always create a runners.yml)
3) Merge `child_image.hex` and `parent_image.hex` to `merged.hex`
4) Split `merged.hex` in `nrfjprog.py` and recreate the two independent hex files we already have in 1) and 2), see https://github.com/nrfconnect/sdk-zephyr/compare
/8038473529fcdbf4fa35f4b37189807a19ad8baf
    - Those are now called: `nrf53_cpuapp.hex` and `nrf53_cpunet.hex` (plus a depfile)
5) Flash the two recreated hex files, `nrf53_cpuapp.hex` and `nrf53_cpunet.hex`.
    (We now have minimum 5 hex files in the build folder)

What I propose here in #324 is to simply use the info we already have, cause the child and parent image always creates independent hex files in each image build folder (and a corresponding runners.yml).
1) Build child_image.hex.
    - and create a child_image/runners.yml
      (each image always create a runners.yml)
2) Build parent_image.hex
    - and create a parent_image/runners.yml
       (each image always create a runners.yml)
3) Use parent_image/runners.yml to flash `parent_image.hex`
     and child_image/runners.yml to flash `child_image.hex`.

Done.

No splitting of hex-files, cause the info already at hand is being used.

We can still take a call 😄

**Comment by oyvindronningstad** [comment]
> > OK, I'm starting to feel a bit lost in all this; could we have a quick call on Monday morning my time / afternoon your time to discuss the problems and solutions? @oyvindronningstad perhaps you could join too?
> 
> Ahh, now I get it, you're talking about this solution compared to #322.
> (I though you were suggesting to extend the info in `runners.yml`)
> 
> The reason for my proposal is to use the info we already have.
> Solution in #322 would do the following:
> 
> 1. Build `child_image.hex`.
>    
>    * and create a child_image/runners.yml
>      (each image always create a runners.yml)
> 2. Build `parent_image.hex`
>    
>    * and create a parent_image/runners.yml
>      (each image always create a runners.yml)
> 3. Merge `child_image.hex` and `parent_image.hex` to `merged.hex`
> 4. Split `merged.hex` in `nrfjprog.py` and recreate the two independent hex files we already have in 1) and 2), see https://github.com/nrfconnect/sdk-zephyr/compare
>    /8038473529fcdbf4fa35f4b37189807a19ad8baf
>    
>    * Those are now called: `nrf53_cpuapp.hex` and `nrf53_cpunet.hex` (plus a depfile)
> 5. Flash the two recreated hex files, `nrf53_cpuapp.hex` and `nrf53_cpunet.hex`.
>    (We now have minimum 5 hex files in the build folder)
> 
> What I propose here in #324 is to simply use the info we already have, cause the child and parent image always creates independent hex files in each image build folder (and a corresponding runners.yml).
> 
> 1. Build child_image.hex.
>    
>    * and create a child_image/runners.yml
>      (each image always create a runners.yml)
> 2. Build parent_image.hex
>    
>    * and create a parent_image/runners.yml
>      (each image always create a runners.yml)
> 3. Use parent_image/runners.yml to flash `parent_image.hex`
>    and child_image/runners.yml to flash `child_image.hex`.
> 
> Done.
> 
> No splitting of hex-files, cause the info already at hand is being used.
> 
> We can still take a call 😄

@tejlmand Please remember that I updated the PR to just do checking for UICR contents.

**Comment by tejlmand** [comment]
> 
> @tejlmand Please remember that I updated the PR to just do checking for UICR contents.

@oyvindronningstad  yes, but I think that is different from the merging and splitting of the hex files.
If I understand the comment from @mbolivar-nordic correctly, then he wants to know the reason for supporting multiple `runners.yml`, instead of the splitting of hex-files.

Secondly, this PR now includes a suggestion to pass the uicr info from CMake here:
https://github.com/nrfconnect/sdk-zephyr/pull/324/commits/8842b771ce53219212d6a7dad1246f71e27b2f2e

**Comment by tejlmand** [comment]
Upstream PR: https://github.com/zephyrproject-rtos/zephyr/pull/26365

**Comment by tejlmand** [comment]
@oyvindronningstad  @mbolivar-nordic  uicr commit removed as agreed.

**Comment by tejlmand** [comment]
> All commits here are [nrf fromlist].

Fixed.

**Comment by carlescufi** [comment]
@mbolivar-nordic can you please merge this as soon as the current upmerge is completed? @utzig requires this commit for his work.

**Comment by utzig** [comment]
> @mbolivar-nordic can you please merge this as soon as the current upmerge is completed? @utzig requires this commit for his work.

Actually I would prefer this PR to be fixed (for internal CI reasons)

**Comment by utzig** [comment]
@mbolivar-nordic Please don't merge this as of yet as this PR is good testing ground for an issue fix; it just passed on my fork/branch of the CI, and as soon as the CI PR is merged upstream I will re-run on master CI!

**Comment by utzig** [comment]
@mbolivar-nordic Feel free to merge!

**Comment by mbolivar-nordic** [comment]
> @mbolivar-nordic Feel free to merge!

Mergeup is done now -- please rebase to resolve conflicts.

**Comment by hakonfam** [comment]
Is this ready to be merged?

**Comment by mbolivar-nordic** [comment]
> Is this ready to be merged?

still needs rebase

**Comment by tejlmand** [comment]
Part of last upmerge.

**Pull request closed without merging**